src/eos-updater-fetch.c

```diff
@@ -690,6 +701,181 @@ prepare_flatpaks_to_deploy (OstreeRepo    *repo,
   return TRUE;
 }
 
+static void
+download_now_cb (GObject    *obj,
+                 GParamSpec *pspec,
+                 gpointer    user_data)
+{
+  MwscScheduleEntry *schedule_entry = MWSC_SCHEDULE_ENTRY (obj);
+  gboolean *download_now_out = user_data;
+  *download_now_out = mwsc_schedule_entry_get_download_now (schedule_entry);
+}
+
+static gboolean
+timeout_cb (gpointer user_data)
+{
+  gboolean *timed_out_out = user_data;
+  *timed_out_out = TRUE;
+  return G_SOURCE_CONTINUE;
+}
+
+/* Create a schedule entry for this download in the download scheduler. */
+static gboolean
+schedule_download (FetchData     *fetch_data,
+                   GMainContext  *context,
+                   GCancellable  *cancellable,
+                   GError       **error)
+{
+  g_autoptr(MwscScheduler) scheduler = NULL;
+  g_autoptr(GAsyncResult) new_result = NULL;
+  g_autoptr(GAsyncResult) schedule_result = NULL;
+  g_auto(GVariantDict) parameters_dict = G_VARIANT_DICT_INIT (NULL);
+  g_autoptr(GVariant) parameters = NULL;
+
+  /* Connect to the download scheduler. If we can’t connect, fail
+   * safe, and don’t download on a potentially metered connection. */
+  mwsc_scheduler_new_async (cancellable, async_result_cb, &new_result);
+
+  while (new_result == NULL)
+    g_main_context_iteration (context, TRUE);
+
+  scheduler = mwsc_scheduler_new_finish (new_result, error);
+
+  if (scheduler == NULL)
+    return FALSE;
+
+  /* Schedule the download. */
+  g_variant_dict_insert (&parameters_dict, "resumable", "b", FALSE);
+  parameters = g_variant_ref_sink (g_variant_dict_end (&parameters_dict));
+
+  mwsc_scheduler_schedule_async (scheduler, parameters, cancellable,
+                                 async_result_cb, &schedule_result);
+
+  while (schedule_result == NULL)
+    g_main_context_iteration (context, TRUE);
```

Comment:
I guess the assumption in this loop, and the loop above, is that they will in practice return quickly (and if Mogwai is buggy and never returns, the bus will time them out after 25 seconds) so you do not need the user-provided timeout to apply here?

Reply:
I had actually neglected to think about timeouts there, but your suggestion is good. The default D-Bus timeouts should work here (we never really expect these paths to fail at all, unless `mogwai-scheduled` fails to start entirely). `scheduling_timeout_seconds` can be kept as the timeout waiting for the signal, which is the more common place to wait.